### PR TITLE
Site setting simplification

### PR DESCRIPTION
**Pull request by hwindsor**
## Pull Request Overview

- Significant refactoring of site settings. Code separated to hopefully be clearer about which parts are for houston internal calls, which parts are for REST GET/POST/PATCH of configuration and which are for GET of the definitions for the REST API.
- Most functionality moved into models code (as having it split between resources and models was causing confusion)
- Only support block GET and get of individual houston only settings (sentryDsn is the only one used). The frontend never GETs individual EDM settings.
- Only support GET of all of the definitions, not individual ones. Frontend never used them.
- Only report the values for the configuration data from edm and houston, not the id and version fields that EDM generates that we don't want to have to emulate when EDM is retired.
- Remove 'currentValue' from the definitions returned data, it was never used by the front end and was an oddity to generate in the houston settings code.
- Removed a couple more 'success' fields from the response, the Front End never uses any of them

DO NOT MERGE YET. This may break things at the moment and needs a bit more testing from Ben

```
GET /api/v1/site-settings/definition/main/block (only one field shown)
Before:
            "site.general.taglineSubtitle": {
                "schema": {
                    "displayType": "longstring",
                    "displayOrder": 1
                },
                "defaultValue": "AI for the conservation of <your species here>.",
                "readOnly": false,
                "isPrivate": false,
                "required": false,
                "displayType": "longstring",
                "settable": true,
                "descriptionId": "CONFIGURATION_SITE_GENERAL_TAGLINESUBTITLE_DESCRIPTION",
                "translationId": "CONFIGURATION_SITE_GENERAL_TAGLINESUBTITLE",
                "labelId": "CONFIGURATION_SITE_GENERAL_TAGLINESUBTITLE_LABEL",
                "name": "configuration_site_general_taglineSubtitle",
                "parentConfigurationId": "site.general",
                "configurationId": "site.general.taglineSubtitle",
                "fieldType": "string",
                "currentValue": "AI for the conservation of zebra."
            },
            
After:
         "site.general.taglineSubtitle": {
                "schema": {
                    "displayType": "longstring",
                    "displayOrder": 1
                },
                "defaultValue": "AI for the conservation of <your species here>.",
                "readOnly": false,
                "isPrivate": false,
                "required": false,
                "displayType": "longstring",
                "settable": true,
                "descriptionId": "CONFIGURATION_SITE_GENERAL_TAGLINESUBTITLE_DESCRIPTION",
                "translationId": "CONFIGURATION_SITE_GENERAL_TAGLINESUBTITLE",
                "labelId": "CONFIGURATION_SITE_GENERAL_TAGLINESUBTITLE_LABEL",
                "name": "configuration_site_general_taglineSubtitle",
                "parentConfigurationId": "site.general",
                "configurationId": "site.general.taglineSubtitle",
                "fieldType": "string"
            },
 ```
 currentValue removed
 
 ```
 GET /api/v1/site-settings/main/block (only one field shown)
Before:
           "site.general.donationButtonUrl": {
                "id": "site.general.donationButtonUrl",
                "value": "",
                "version": 1656488503307
            }, 
After:
            "site.general.donationButtonUrl": {
                "value": ""
            },
```
id and version removed

### CODE REVIEW

File: tests/modules/site_settings/resources/test_file_operations.py [hunk 1]
```diff
@@ -53,10 +53,7 @@ def test_file_settings(admin_user, flask_app_client, flask_app, db, request, tes
         site_setting = resp.json
         assert site_setting['key'] == 'footer_image'
 
-        # this is to test that GET /api/v1/site-settings/file is *truly* only getting file-type values
```
Comment (hwindsor) [inline tests/modules/site_settings/resources/test_file_operations.py:56]: Can no longer set arbitrary houston keys, they must be in HOUSTON SETTINGS but there are plenty of them so this isn't necessary

File: app/modules/site_settings/models.py [hunk 2]
```diff
@@ -23,14 +23,13 @@ class SiteSetting(db.Model, Timestamp):
...
         'confirm_deleted_rows': False,
     }
-    EDM_PREFIX = 'site.'
 
     # public=False means the setting is only used in the backend
     HOUSTON_SETTINGS = {
         'email_service': {
             'type': str,
             'public': False,
-            'edm_definition': {
```
Comment (hwindsor) [inline app/modules/site_settings/models.py:33]: anything that says edm_ is being removed, this is no longer the edm format, this is the format we will be using going forward.

File: app/modules/site_settings/models.py [hunk 3]
```diff
@@ -517,37 +476,265 @@ def get_edm_configuration(cls, key, default=None, **kwargs):
...
+                    species_json['suggestedValues'].insert(
+                        0,
+                        {
+                            'scientificName': scientific_name,
+                            'commonNames': [details.get('common_name', scientific_name)],
+                            'itisTsn': details.get('itis_id'),
+                        },
+                    )
+
+        # TODO also traverse private here FIXME
```
Comment (hwindsor) [inline app/modules/site_settings/models.py:699]: @naknomum, This was a hanging comment in the resources file. Is this still relevant?

File: app/modules/site_settings/models.py [hunk 4]
```diff
@@ -621,7 +812,7 @@ def __init__(self, *args, **kwargs):
         else:
             from app.modules.site_settings.models import SiteSetting
 
-            data = SiteSetting.get_edm_configuration('site.custom.regions')
```
Comment (hwindsor) [inline app/modules/site_settings/models.py:624]: should never have known it was reading an edm value anyway

File: tests/modules/site_settings/resources/test_bundles.py [hunk 5]
```diff
@@ -81,7 +80,6 @@ def test_bundle_read(
         ({'email_default_sender_name': 'testing'}, 200, False),
         (
             {
-                'bad_key': 'abcd',
```
Comment (hwindsor) [inline tests/modules/site_settings/resources/test_bundles.py:84]: @naknomum This is the test I'm really not sure about so please go through this with a fine toothcomb assuming that I've made a mistake somewhere :)

File: app/modules/site_settings/models.py [hunk 6]
```diff
@@ -94,11 +93,21 @@ class SiteSetting(db.Model, Timestamp):
             'public': False,
             'default': 'Wildbook is copyrighted 2022 by Wild Me.  Individual contributors to Wildbook retain copyrights for submitted images. No reproduction or usage of material in this library is permitted without the express permission of the copyright holders.',
         },
-        'social_group_roles': {'type': list, 'public': True},
+        'social_group_roles': {
+            'type': list,
+            'public': True,
+            'validate_function': lambda value: SiteSetting.validate_social_group_roles(
```
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/site_settings/models.py:99]: You must have been looking at an earlier code drop, the validate_function is used now.

File: app/modules/site_settings/models.py [hunk 7]
```diff
@@ -218,10 +227,13 @@ def set(
         boolean=None,
         override_readonly=False,
     ):
-        if is_extension_enabled('edm') and cls.is_edm_key(key):
-            raise ValueError(
-                f'forbidden to directly set key with prefix "{cls.EDM_PREFIX}" via (key={key})'
-            )
+        if (
```
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/site_settings/models.py:230]: This is certainly possible but I cut off at a point where it was a lot clearer than it was.

File: app/modules/site_settings/resources.py [hunk 8]
```diff
@@ -173,155 +172,35 @@ class MainConfigurationDefinition(Resource):
...
-        if SiteSetting.is_block_key(path):
-            data = self._insert_houston_definitions(data)
-
-        # TODO also traverse private here FIXME
-        return data
+        # Currently only support getting all definitions as one
+        if SiteSetting.is_rest_block_key(path):
+            return SiteSetting.get_all_rest_definitions()
+        else:
+            abort(400, 'Can only get the full block of definitions, not individual items')
```
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/site_settings/resources.py:180]: I need to talk to Ben about this. If you look at how this is used in the frontend, I think there needs to be a ew API that does what the FE wants in one message rather than 2

File: app/modules/site_settings/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (karenc) [reply app/modules/site_settings/models.py:230]: We can certainly do this in a later PR.  There's validation in `set_key_value` but not in `set`.  Just looking at the method names, it's not really easy to know which one to use

File: app/modules/site_settings/resources.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (karenc) [reply app/modules/site_settings/resources.py:180]: I'm suggesting:

```python
@api.route('/definition/main/block')
class MainConfigurationDefinition(Resource):
    def get(self):
        return SiteSetting.get_all_rest_definitions()
```

The API would return 400 anyway for anything else, right?

[the remark this answers is not in the document]
Comment (karenc) [reply app/modules/site_settings/resources.py:180]: Oh I guess we need to add `@api.route('/definition/main/__bundle_setup')`

**Review by brmscheiner — APPROVED** [review]
did some testing, not noticing any affects on the FE

File: app/modules/site_settings/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/site_settings/models.py:99]: I actually tried to use `'validate_function': SiteSetting.validate_social_group_roles` but that didn't work as it requires the SiteSetting class to exist and I think that at this point the class is being created (at least pycharm was reporting it as an error and this matches my experience of trtying similar things in C++) so I used the lambda instead. 
Regarding where the validate_ functions should go, I can see the argument that it  could be moved anywhere but the advantage of having them all in SiteSettings is that it does not require the set_key_value to import the required module. 

TODO removal, good point

File: app/modules/site_settings/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/site_settings/models.py:230]: I definitely see what you mean, you have a valid point. I hadn't spotted that. 
This change was to tidy up the rats nest of REST related APIs and for the most part I'd left the 'internal houston to site settings' apis alone. 
I'm keen to not add too much more to this PR, it's virtually a complete rewrite of the module as it is but now you've made me aware of this, it may get fixed 'quite soon'.
Also getting this PR in will unblock both EDM retirement and autogenerated names so I'd rather not polish this any more right now.

File: app/modules/site_settings/models.py [hunk 9]
```diff
@@ -428,19 +398,12 @@ def get_boolean(cls, key, default=None):
...
 
-    # a bit of hackery.  right now *all* keys in edm-configuration are of the form `site.foo` so we use
-    #   as a way branch on _where_ to get the value to return here.  but as we ween ourselves off edm config,
-    #   either gradually or forklift, this can be extended to allow that
-    @classmethod
-    def is_edm_key(cls, key):
-        return key.startswith(cls.EDM_PREFIX)
-
     @classmethod
     def get_value(cls, key, default=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/site_settings/models.py:402]: Again you have a valid point and yes, this might get revisited 'real soon' but I'd like to get this in as is.

File: app/modules/site_settings/resources.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hwindsor) [reply app/modules/site_settings/resources.py:180]: I do see what you are saying but yeah, it is possible to read the block via `/definition/main/` `/definition/main/block` and `/definition/main/_bundle_setup`
But as I say, the goal is to combine the block configuration and the definition into one message at which point the existing definition and block configuration endpoints will be rertired so I'm not keen on optimising them.

File: tests/modules/site_settings/resources/test_bundles.py (hunk 5 above)
Comment (karenc) [reply tests/modules/site_settings/resources/test_bundles.py:84]: I was the last one to touch this test so I had a look at it.  The API has changed so partial update doesn't work anymore.  I think the following changes to the test makes more sense to the current code:

```diff
@@ -74,45 +74,63 @@ def test_bundle_read(
     module_unavailable('site_settings'), reason='Site-Settings module disabled'
 )
 @pytest.mark.parametrize(
-    'data,expected_status_code,is_edm',
+    'data,invalid_keys, expected_status_code,is_edm',
     (
-        ({'site.name': f'TEST-{str(uuid.uuid4())}'}, 200, True),
-        ({'email_default_sender_name': 'testing'}, 200, False),
+        # Case 1: One valid site setting (succeeds)
+        ({'site.name': f'TEST-{str(uuid.uuid4())}'}, [], 200, True),
+        # Case 2: More than one valid site settings (succeeds)
         (
             {
+                'email_default_sender_name': 'testing',
+                'recaptchaPublicKey': 'recaptcha-public-key',
+            },
+            [],
+            200,
+            False,
+        ),
+        # Case 3: One invalid key and two invalid types (fails)
+        (
+            {
+                'bad_key': 'abcd',
                 'email_default_sender_name': {'foo': 'bar'},
                 'email_default_sender_email': [],
             },
+            ['bad_key', 'email_default_name', 'email_default_sender_name'],
             400,
             False,
         ),
+        # Case 4: Two valid site settings and one invalid key (fails)
         (
             {
                 'email_default_sender_name': 'Testing',
                 'bad_key': 'abcd',
                 'site.name': f'TEST-{str(uuid.uuid4())}',
             },
+            ['bad_key'],
             400,
             True,
         ),
-        ({'recaptchaPublicKey': 'recaptcha-public-key'}, 200, False),
     ),
 )
 def test_bundle_modify(
-    flask_app_client, admin_user, db, request, data, expected_status_code, is_edm
+    flask_app_client,
+    admin_user,
+    db,
+    request,
+    data,
+    invalid_keys,
+    expected_status_code,
+    is_edm,
 ):
     if extension_unavailable('edm') and is_edm:
         pytest.skip('EDM extension disabled')
 
     response = conf_utils.read_main_settings(flask_app_client, admin_user)
     old_values = {}
-    invalid_key = []
     for key in data:
         old_value = response.json['response']['configuration'].get(key, {})
         if old_value:
             old_values[key] = old_value['value']
-        else:
-            invalid_key.append(key)
     request.addfinalizer(
         lambda: conf_utils.modify_main_settings(flask_app_client, admin_user, old_values)
     )
@@ -126,14 +144,13 @@ def test_bundle_modify(
         assert response.json['updated']
         assert isinstance(response.json['updated'], list)
         for key, value in data.items():
-            if key in invalid_key:
-                assert key not in response.json['updated']
-                assert SiteSetting.get_value(key) is None
-            else:
-                assert key in response.json['updated']
-                assert SiteSetting.get_value(key) == value
+            assert key in response.json['updated']
+            assert SiteSetting.get_value(key) == value
     else:
-        for key in data:
-            if key in invalid_key:
-                assert key in response.json['message']
-                break
+        for key in invalid_keys:
+            if key in response.json['message']:
+                pass  # break
+        else:
+            assert (
+                False
+            ), f'Expected one of {invalid_keys} to be in message "{response.json["message"]}"'
```

File: tests/modules/site_settings/resources/test_read_configurations.py [hunk 10]
```diff
@@ -205,7 +178,7 @@ def test_alter_edm_custom_fields(flask_app_client, admin_user):
     patch_data = [
         {
             'op': 'remove',
-            'path': f'/site.custom.customFields.Occurrence/{first["id"]}',
```
[the remark this answers is not in the document]
Comment (karenc) [reply tests/modules/site_settings/resources/test_read_configurations.py:208]: I looked into this test and I think what you've done is correct.

With the leading slash, I get a 400 response like:

```json
{"status": 400,
 "message": "removal of /site.custom.customFields.Occurrence/d0370568-8fe7-4637-a6c6-71eb843daef8 not supported"}
```

Also it seems `site.custom.customFields.Occurrence` and `site.custom.customFields.Sighting` are both in sync and so we only need to remove a field from one of them to remove from both.

File: app/modules/site_settings/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (naknomum) [reply app/modules/site_settings/models.py:99]: i do like the idea of the validate_function option here.  will definitely help with movement of stuff from edm.

File: app/modules/site_settings/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (naknomum) [reply app/modules/site_settings/models.py:400]: i was considering this exact thing when i added the `get_boolean` ... yes lets revisit.  i probably ran into the same small edge cases that made me balk on it.  but i think it would be desirable, especially when we know we can trust the _type_ property to hint us on what to actually get.

File: tests/modules/site_settings/resources/test_bundles.py (hunk 5 above)
Comment (hwindsor) [reply tests/modules/site_settings/resources/test_bundles.py:84]: I agree, that is a lot more readable. If you have this on a branch somewhere, I would definitely agree with it going into main.

Comment (karenc) [reply tests/modules/site_settings/resources/test_bundles.py:84]: https://github.com/WildMeOrg/houston/pull/738
